First cut of seismic waveform example

ptomecek opened the pull request:
It's in a reasonably presentable state, but could always use further improvements.

robambalu commented on examples/07_end_to_end/seismic_waveform.ipynb:
```diff
@@ -0,0 +1,1334 @@
+     "name": "stderr",
+     "output_type": "stream",
+     "text": [
+      "Exception in thread Thread-7:\n",
```
dont think you meant to leave this here

robambalu commented on examples/07_end_to_end/seismic_waveform.ipynb:
```diff
@@ -0,0 +1,1334 @@
+    "        parsed_url = urlparse(server_url, scheme=\"seedlink\")\n",
+    "        self.server_hostname = parsed_url.hostname\n",
+    "        self.server_port = parsed_url.port or 18000\n",
+    "        self.conn = SeedLinkConnection()\n",
```
may want to move this into the impl and just pass server_url to the impl
better demonstration that the wiring time object should really just hold the necessary args

robambalu commented on examples/07_end_to_end/seismic_waveform.ipynb:
```diff
@@ -0,0 +1,1334 @@
+    "    def __init__(self, engine, conn: SeedLinkConnection):\n",
+    "        super().__init__(engine)\n",
+    "\n",
+    "        self.conn = conn\n",
```
any reason this isnt "private"?

ptomecek commented:
@robambalu Latest commit should incorporate all your feedback

AdamGlustein commented on examples/07_end_to_end/seismic_waveform.ipynb:
```diff
@@ -0,0 +1,1322 @@
+    "from typing import Dict, List, Optional, Tuple\n",
+    "\n",
+    "warnings.filterwarnings('ignore') "
+   ]
```
I'd put this code after the introduction and make a note that the user must `pip install obspy` as it's not a required csp dependency. Additionally, `scipy` isn't a required csp dependency either.

AdamGlustein commented on examples/07_end_to_end/seismic_waveform.ipynb:
```diff
@@ -0,0 +1,1322 @@
+   "metadata": {},
+   "source": [
+    "## Simulation\n",
+    "In real-time, trace data will be sent in batches of observations, so we can use csp to simulate this by writing a node that divides a stream into a time series of pieces based on an interval, and assuming a certain transmission lag"
```
I found this sentence confusing. I would change:
"divides a stream into a time series of pieces based on an interval, and assuming a certain transmission lag" 
to 
"divides a stream into fixed sized chunks for each interval. We can also simulate a lag in the transmitted data, as we would see in realtime event processing."

AdamGlustein commented on examples/07_end_to_end/seismic_waveform.ipynb:
```diff
@@ -0,0 +1,1322 @@
+   "id": "199cc6f0-7103-4d93-8e4b-ffbf17d07200",
+   "metadata": {},
+   "source": [
+    "We can now run this simple graph to show how the pieces are divided into 5 second chunks with a small simulated transmission delay"
```
I would make a note in this text block why one sees 3 data chunks per interval (it's the 3 coordinate axes, N, E and Z).
```
2009-08-24 00:20:09 Received simulated trace BW.RJOB..EHZ:BW.RJOB..EHZ | 2009-08-24T00:20:03.000000Z - 2009-08-24T00:20:08.990000Z | 100.0 Hz, 600 samples
2009-08-24 00:20:09 Received simulated trace BW.RJOB..EHN:BW.RJOB..EHN | 2009-08-24T00:20:03.000000Z - 2009-08-24T00:20:08.990000Z | 100.0 Hz, 600 samples
2009-08-24 00:20:09 Received simulated trace BW.RJOB..EHE:BW.RJOB..EHE | 2009-08-24T00:20:03.000000Z - 2009-08-24T00:20:08.990000Z | 100.0 Hz, 600 samples
```

AdamGlustein commented on examples/07_end_to_end/seismic_waveform.ipynb:
```diff
@@ -0,0 +1,1322 @@
+    "## Analytics\n",
+    "\n",
+    "### Filtering\n",
+    "With waveform data incoming in chunks (either historically or real time), we can use csp to implement filtering logic leveraging scipy. Here we illustrate a simple bandpass filter, but this could easily be generalized to other filter types.\n",
```
"With waveform data coming in chunks in both historical and realtime runs"

AdamGlustein commented on examples/07_end_to_end/seismic_waveform.ipynb:
```diff
@@ -0,0 +1,1322 @@
+   "execution_count": null,
+   "id": "2ff30a46-93eb-4599-bd53-9246c3c6800d",
+   "metadata": {},
+   "outputs": [],
```
empty code block here

AdamGlustein commented on examples/07_end_to_end/seismic_waveform.ipynb:
```diff
@@ -0,0 +1,1322 @@
+   "id": "e9159847-c974-4e0a-a0f6-66de8b5c3f2e",
+   "metadata": {},
+   "source": [
+    "Is is now very straightforward to write the actual triggering logic based on the characteristic function in csp; logic that will work just as well in real time as on historical data"
```
typo: "It is"

AdamGlustein commented on examples/07_end_to_end/seismic_waveform.ipynb:
```diff
@@ -0,0 +1,1322 @@
+   "source": [
+    "The \"id\" of the Trace object is composed of the network id (\"BW\"), the station id (\"RJOB\") and the channel (\"EHZ\"). The channel names are composed of the band code (i.e. frequency, where \"E\" is \"Extremely High Frequency\"), the instrument code (\"H\" is \"High Gain Seismometer\") and the orientation code (i.e. \"Z\" is vertical movement). You can read more about channel naming [he ...
+    "\n",
+    "The trace has two main attributes, \"stats\" which contains metadata such as the start/end time, sampling frequency, etc and \"data\" which is a numpy array of the raw measurements, at equally sampled intervals in the time range."
```
What actually are the raw measurements? Positional deviations? Also what units are these measurements in. 
It would also be interesting if you alluded to how these measurements can be used for insights into tremors/earthquakes. We wouldn't do any of that in the example, but I think it would add more interest to the notebook.

ptomecek replied:
There is actually a bit of an unanswered units question as the real time and historical units don't seem to align, and I don't know the reason. Maybe @melissawm can reach out to some obspy folks to understand that better. This is good feedback though, and we can put more of that info in the introduction.

AdamGlustein commented on examples/07_end_to_end/seismic_waveform.ipynb:
```diff
@@ -0,0 +1,1322 @@
+   "metadata": {},
+   "source": [
+    "For this example, we use a station on the Canadian (i.e. \"CN\") network, as the high frequency channel will publish updates every 4-5 seconds, which is more rapid than the IRIS network used above. However, occasionally the connection to this network can time out, in which case, please try again later. When it's working properly, a log line containing `INFO: network socket opened` should app ...
+    "The [GEOFON network](https://geofon.gfz-potsdam.de/waveform/seedlink.php) (http://geofon.gfz-potsdam.de:18000) is also a good alternative choice, i.e. channel (\"GE\", \"STU\", \"HHZ\"). "
```
The Canadian network was timed out when I tried, but I used the German one and it worked.
Maybe we should just use the German one for this example and mention the Canadian one as an alternative for people top try out?

ptomecek replied:
Not allowed to for compliance reasons.

ptomecek commented:
@AdamGlustein All your other feedback is incorporated, but I would also love an answer to the units question, so leaving that one open for now.

timkpaine reviewed (changes requested):
`EasySeedLinkAdapter` is pretty useless on its own, i would combine with `EasySeedLinkAdapterImpl` as [here](https://github.com/Point72/csp/blob/60e447340ef8897838e8a5fae3a1615fd9572345/examples/04_writing_adapters/e7_adaptermanager_inputoutput.py#L35). 

Also `EasySeedLinkAdapterImpl` / `_EasySeedLinkAdapterImpl` , `EasySeedLinkAdapter` / `_EasySeedLinkAdapter` is pretty awful naming unless the goal is to confuse the reader 😆

ptomecek commented:
> `EasySeedLinkAdapter` is pretty useless on its own, i would combine with `EasySeedLinkAdapterImpl` as [here](https://github.com/Point72/csp/blob/60e447340ef8897838e8a5fae3a1615fd9572345/examples/04_writing_adapters/e7_adaptermanager_inputoutput.py#L35).
> 
> Also `EasySeedLinkAdapterImpl` / `_EasySeedLinkAdapterImpl` , `EasySeedLinkAdapter` / `_EasySeedLinkAdapter` is pretty awful naming unless the goal is to confuse the reader 😆

I will have a look at this - I probably copied the wrong example

ptomecek commented:
@timkpaine I've updated the adapter definition - agree it's much more straightforward now. 
@AdamGlustein I've added a link to the USGS FAQ on seismography that has more background information about the space for lay people.

timkpaine reviewed (dismissed):
- [ ] Needs tests
- [ ] `README.md` in `07_end_to_end` similar to [this](https://github.com/Point72/csp/blob/827558828e572ce1ae8a7c91de262652e79fa022/examples/06_advanced/README.md)
- [ ] Outputs stripped (meh) or notebook stored in LFS (also meh)

ptomecek commented:
> * [ ]  Needs tests
> * [ ]  `README.md` in `07_end_to_end` similar to [this](https://github.com/Point72/csp/blob/827558828e572ce1ae8a7c91de262652e79fa022/examples/06_advanced/README.md)
> * [ ]  Outputs stripped (meh) or notebook stored in LFS (also meh)

As discussed offline, always keen on testing, let's discuss strategies next week. 
I will also follow up on the longer term question about binary assets in the repo and possible move to LFS or separate repo for binary-heavy examples/notebooks.

robambalu commented on examples/07_end_to_end/seismic_waveform.ipynb:
```diff
@@ -0,0 +1,1325 @@
+    "        self._starttime = None\n",
+    "        self._endtime = None\n",
+    "\n",
+    "    def subscribe(self, network, station, selector=None, push_mode=csp.PushMode.NON_COLLAPSING):\n",
```
this isnt the correct way to expose adapter managers.  This is the impl ( runtime ) instance that should only be created by the engine when it starts running.
It shouldnt have the user facing subscribe API, which should be a separate light weight "graph time" object

ptomecek replied:
I changed this back,

robambalu commented on examples/07_end_to_end/seismic_waveform.ipynb:
```diff
@@ -0,0 +1,1325 @@
+    "        return _easy_seedlink_adapter(self, network, station, selector, push_mode=push_mode)\n",
+    "\n",
+    "    def _create(self, engine, memo):\n",
+    "        # We'll avoid having a second class and make our AdapterManager and AdapterManagerImpl the same\n",
```
I see youre aware of this but its bad form, I dont think we should be showcasing bad form in our examples

robambalu commented on examples/07_end_to_end/seismic_waveform.ipynb:
```diff
@@ -0,0 +1,1325 @@
+    "from csp.impl.wiring import py_push_adapter_def\n",
+    "\n",
+    "\n",
+    "class EasySeedLinkAdapter(AdapterManagerImpl):\n",
```
May want to call this EasySeedLinkAdapterManager for clarity

ptomecek commented:
> * [ ]  Needs tests
> * [ ]  `README.md` in `07_end_to_end` similar to [this](https://github.com/Point72/csp/blob/827558828e572ce1ae8a7c91de262652e79fa022/examples/06_advanced/README.md)
> * [ ]  Outputs stripped (meh) or notebook stored in LFS (also meh)

As discussed, we will work on tests separately. I think we can also put together a README for the section in a separate PR.
